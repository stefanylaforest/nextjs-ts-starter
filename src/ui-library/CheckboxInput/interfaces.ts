export default interface CheckboxInputProps {
  name: string;
  label: string | React.ReactNode;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  checked: boolean;
  disabled?: boolean;
}
