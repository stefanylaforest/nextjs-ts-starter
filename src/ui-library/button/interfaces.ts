type ButtonProps = {
  children: React.ReactNode;
  onClick: () => void;
  type?: 'submit' | 'reset' | 'button';
  disabled?: boolean;
  size?: 'large' | 'small';
} & ({ variant: 'primary' | 'secondary' } | { variant: 'icon'; ariaLabel: string });

export type Ref = HTMLButtonElement;

export default ButtonProps;
