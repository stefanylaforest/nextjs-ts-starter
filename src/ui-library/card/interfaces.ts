type RequireAtLeastOne<T> = {
  [K in keyof T]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<keyof T, K>>>;
}[keyof T];

interface AvailableCardProps {
  header: string | React.ReactNode;
  body: string | React.ReactNode;
  footer: string | React.ReactNode;
}

type CardProps = RequireAtLeastOne<AvailableCardProps>;

export default CardProps;
