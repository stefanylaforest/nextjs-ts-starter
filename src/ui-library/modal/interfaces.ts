export default interface ModalProps {
  show: boolean;
  closeModal: () => void;
  title?: string;
  children: React.ReactNode;
  footer?: React.ReactNode;
}
