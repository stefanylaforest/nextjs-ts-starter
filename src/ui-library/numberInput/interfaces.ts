export default interface NumberInputProps {
  label: string;
  name: string;
  value: number | string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  required?: boolean;
  min?: number;
  max?: number;
  readonly?: boolean;
  step?: number;
}
