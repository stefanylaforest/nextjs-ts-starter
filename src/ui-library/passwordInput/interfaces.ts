type PasswordInputProps =
  | {
      value: string;
      name: string;
      onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
      placeholder?: string;
    } & (
      | { label?: undefined; requiredSymbol?: never }
      | { label: string; requiredSymbol?: boolean }
    );

export default PasswordInputProps;
