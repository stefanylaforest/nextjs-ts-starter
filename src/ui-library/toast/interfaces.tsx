export interface BasicToast {
  message: string | React.ReactNode;
  type: 'info' | 'success' | 'error' | 'warning';
}

export interface ToastWithId extends BasicToast {
  id: string;
}

export interface ToastContextType {
  activate: ({ message, type }: BasicToast) => void;
}

export default interface ToastProps extends ToastWithId {
  remove: (key: string) => void;
}
