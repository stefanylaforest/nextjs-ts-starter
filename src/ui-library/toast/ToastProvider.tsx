import React, { createContext, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';

import Toast from './toast';
import styles from './toast.module.scss';
import { ToastWithId, ToastContextType, BasicToast, ToastProviderProps } from './interfaces';

const ToastContext = createContext<ToastContextType | null>(null);

const ToastProvider = ({ children }: ToastProviderProps) => {
  const [toasts, setToasts] = useState<ToastWithId[]>([]);

  const activate = ({ message, type }: BasicToast) => {
    setToasts([...toasts, { id: uuidv4(), message, type }]);
  };

  const removeToast = (id: string) => {
    const filteredToasts = toasts.filter((toast) => toast.id !== id);
    setToasts(filteredToasts);
  };

  return (
    <ToastContext.Provider value={{ activate }}>
      {children}
      <div className={styles.toastWrapper}>
        {toasts.map((toast) => (
          <Toast
            key={toast.id}
            type={toast.type}
            message={toast.message}
            remove={removeToast}
            id={toast.id}
          />
        ))}
      </div>
    </ToastContext.Provider>
  );
};

export { ToastContext, ToastProvider };
