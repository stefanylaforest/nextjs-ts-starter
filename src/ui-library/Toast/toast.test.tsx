import React, { useContext, useEffect } from 'react';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import { ToastProvider, ToastContext } from './ToastProvider';
import { ToastContextType } from './interfaces';

interface Type {
  type: 'info' | 'success' | 'error' | 'warning';
}

const TestComponent = ({ type = 'info' }: Type) => {
  const toast = useContext(ToastContext) as ToastContextType;
  useEffect(() => {
    toast.activate({
      message: 'toast message',
      type: type,
    });
  }, []);
  return (
    <div>
      <p>test</p>
    </div>
  );
};

describe('toast component', () => {
  it('it should fire a toast', () => {
    render(
      <ToastProvider>
        <TestComponent type={'info'} />
      </ToastProvider>
    );
    expect(screen.getByRole('alert', { name: '' })).toBeInTheDocument();
  });

  it('it should container a message', () => {
    render(
      <ToastProvider>
        <TestComponent type={'info'} />
      </ToastProvider>
    );
    expect(screen.getByRole('alert', { name: '' })).toHaveTextContent('toast message');
  });

  it('it should dissmiss toast on click of close button within toast', async () => {
    render(
      <ToastProvider>
        <TestComponent type={'info'} />
      </ToastProvider>
    );
    const toast = screen.getByRole('alert', { name: '' });
    const closeButton = screen.getByRole('button', { name: 'dismiss' });
    await userEvent.click(closeButton);
    expect(toast).not.toBeInTheDocument();
  });

  it('it should render an info toast', () => {
    const { container } = render(
      <ToastProvider>
        <TestComponent type={'info'} />
      </ToastProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('it should render a success toast', () => {
    const { container } = render(
      <ToastProvider>
        <TestComponent type={'success'} />
      </ToastProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('it should render an error toast', () => {
    const { container } = render(
      <ToastProvider>
        <TestComponent type={'error'} />
      </ToastProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('it should render an warning toast', () => {
    const { container } = render(
      <ToastProvider>
        <TestComponent type={'warning'} />
      </ToastProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('it should dissapear after timer', async () => {
    jest.useFakeTimers();
    render(
      <ToastProvider>
        <TestComponent type={'warning'} />
      </ToastProvider>
    );
    const toast = screen.getByRole('alert', { name: '' });
    act(() => {
      jest.runAllTimers();
    });
    expect(toast).not.toBeInTheDocument();
    jest.useRealTimers();
  });
});
