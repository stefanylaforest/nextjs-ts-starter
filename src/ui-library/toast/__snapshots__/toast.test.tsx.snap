// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toast component it should render a success toast 1`] = `
<div>
  <div>
    <p>
      test
    </p>
  </div>
  <div
    class="toastWrapper"
  >
    <div
      class="container success animate"
      role="alert"
    >
      <div
        class="rightCol"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 12 2 2 4-4m6 2a9 9 0 1 1-18 0 9 9 0 0 1 18 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
         
        toast message
      </div>
      <div
        class="leftCol"
      >
        <button
          aria-label="dismiss"
          class="icon"
          type="button"
        >
          <svg
            class="h-6 w-6"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18 18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`toast component it should render an error toast 1`] = `
<div>
  <div>
    <p>
      test
    </p>
  </div>
  <div
    class="toastWrapper"
  >
    <div
      class="container error animate"
      role="alert"
    >
      <div
        class="rightCol"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 8v4m0 4h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
         
        toast message
      </div>
      <div
        class="leftCol"
      >
        <button
          aria-label="dismiss"
          class="icon"
          type="button"
        >
          <svg
            class="h-6 w-6"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18 18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`toast component it should render an info toast 1`] = `
<div>
  <div>
    <p>
      test
    </p>
  </div>
  <div
    class="toastWrapper"
  >
    <div
      class="container info animate"
      role="alert"
    >
      <div
        class="rightCol"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13 16h-1v-4h-1m1-4h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
         
        toast message
      </div>
      <div
        class="leftCol"
      >
        <button
          aria-label="dismiss"
          class="icon"
          type="button"
        >
          <svg
            class="h-6 w-6"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18 18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`toast component it should render an warning toast 1`] = `
<div>
  <div>
    <p>
      test
    </p>
  </div>
  <div
    class="toastWrapper"
  >
    <div
      class="container warning animate"
      role="alert"
    >
      <div
        class="rightCol"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-3L13.732 4c-.77-1.333-2.694-1.333-3.464 0L3.34 16c-.77 1.333.192 3 1.732 3z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
         
        toast message
      </div>
      <div
        class="leftCol"
      >
        <button
          aria-label="dismiss"
          class="icon"
          type="button"
        >
          <svg
            class="h-6 w-6"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18 18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
